smithy-model/src/main/java/software/amazon/smithy/model/shapes/EnumShape.java

```diff
@@ -0,0 +1,413 @@
+     * or if the enum definitions don't have names.
+     *
+     * @param shape A base {@link StringShape} to convert.
+     * @return Optionally returns an {@link EnumShape} equivalent of the given shape.
+     */
+    public static Optional<EnumShape> fromStringShape(StringShape shape) {
+        return fromStringShape(shape, false);
+    }
+
+    @Override
+    public ShapeType getType() {
+        return ShapeType.ENUM;
+    }
+
+    public static final class Builder extends StringShape.Builder {
+        private final BuilderRef<Map<String, MemberShape>> members = BuilderRef.forOrderedMap();
+
+        @Override
+        public EnumShape build() {
+            addSyntheticEnumTrait();
+            return new EnumShape(this);
+        }
+
+        private void addSyntheticEnumTrait() {
+            SyntheticEnumTrait.Builder builder = SyntheticEnumTrait.builder();
+            for (MemberShape member : members.get().values()) {
+                try {
+                    builder.addEnum(EnumDefinition.fromMember(member));
+                } catch (IllegalStateException e) {
+                    // This can happen if the enum value trait is using something other
+                    // than a string value. Rather than letting the exception propagate
+                    // here, we let the shape validator handle it because it will give
+                    // a much better error.
+                    return;
+                }
+            }
+            addTrait(builder.build());
+        }
+
+        @Override
+        public ShapeType getShapeType() {
+            return ShapeType.ENUM;
+        }
+
+        @Override
+        public Builder id(ShapeId shapeId) {
+            super.id(shapeId);
+            for (MemberShape member : members.peek().values()) {
+                addMember(member.toBuilder().id(shapeId.withMember(member.getMemberName())).build());
+            }
+            return this;
+        }
+
+        public Builder members(EnumTrait trait, boolean synthesizeNames) {
+            if (getId() == null) {
+                throw new IllegalStateException("An id must be set before adding a named enum trait to a string.");
+            }
+            clearMembers();
+
+            for (EnumDefinition definition : trait.getValues()) {
+                Optional<MemberShape> member = definition.asMember(getId(), synthesizeNames);
+                if (member.isPresent()) {
+                    addMember(member.get());
+                } else {
+                    throw new IllegalStateException(String.format(
+                            "Unable to convert enum trait entry with name: `%s` and value `%s` to an enum member.",
+                            definition.getName().orElse(""), definition.getValue()
+                    ));
+                }
+            }
+
+            return this;
+        }
+
+        public Builder members(EnumTrait trait) {
```

Comment:
I don't get what this is doing. Is this actually something like "setMembersFromEnumTrait" or something?